cmd/state-exec/main.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"time"
+
+	"github.com/ActiveState/cli/cmd/state-exec/internal/logr"
+)
+
+const (
+	executorName     = "state-exec"
+	envVarKeyVerbose = "ACTIVESTATE_VERBOSE"
+	userErrMsg       = "Not user serviceable; Please contact support for assistance."
+)
+
+var (
+	logErr = func(format string, args ...interface{}) {
+		fmt.Fprintf(os.Stderr, "%s: ", executorName)
+		fmt.Fprintf(os.Stderr, format+"\n", args...)
+	}
+)
+
+func logDbgFunc(start time.Time) logr.LogFunc {
+	return func(format string, args ...interface{}) {
+		fmt.Fprintf(os.Stderr, "[DEBUG %9d] ", time.Since(start).Nanoseconds())
+		fmt.Fprintf(os.Stderr, format+"\n", args...)
+	}
+}
+
+func init() {
+	runtime.GOMAXPROCS(1)
+}
+
+func main() {
+	if os.Getenv(envVarKeyVerbose) == "true" {
+		logr.SetDebug(logDbgFunc(time.Now()))
+	}
+
+	if err := run(); err != nil {
+		if exitErr := (&exec.ExitError{}); errors.As(err, &exitErr) {
+			os.Exit(exitErr.ExitCode())
+		}
+
+		logErr(userErrMsg)
+		logErr("%s", err)
+		os.Exit(1)
+	}
+
+	os.Exit(0)
+}
+
+func run() error {
+	logr.Debug("hello")
+	defer logr.Debug("run goodbye")
+
+	hb, err := newHeartbeat()
+	if err != nil {
+		return err
+	}
+	logr.Debug("message data - pid: %s, exec: %s", hb.ProcessID, hb.ExecPath)
+
+	meta, err := newExecutorMeta(hb.ExecPath)
```

Comment:
The executor uses its own path to get the meta data so that it can know how to contact the service (sockpath). Included in that data is the project name and organization. We could, instead, send the same location through the socket and have the service lookup (from disk or cache), but it's quite trivial for it to just be done in one type (and, therefore, one place). In other words, splitting it up will make it more complex and also require more processing.